backend/src/main/scala/cromwell/backend/validation/ContinueOnReturnCodeValidation.scala

```diff
@@ -0,0 +1,46 @@
+package cromwell.backend.validation
+
+import cromwell.backend.validation.RuntimeAttributesValidation._
+import cromwell.core._
+import wdl4s.types.{WdlArrayType, WdlStringType}
+import wdl4s.values.{WdlArray, WdlBoolean, WdlInteger, WdlString}
+
+import scalaz.Scalaz._
+
+trait ContinueOnReturnCodeValidation extends RuntimeAttributesValidation[ContinueOnReturnCode] {
+  override def key = RuntimeAttributesKeys.ContinueOnReturnCodeKey
+
+  override def staticDefaultOption = Option(WdlInteger(0))
+
+  override def coercion = ContinueOnReturnCode.validWdlTypes
+
+  override protected def failureMessage = {
+    s"Expecting $key runtime attribute to be either a Boolean, a String 'true' or 'false', or an Array[Int]"
+  }
+
+  override def validateValue = {
+    case WdlBoolean(value) => ContinueOnReturnCodeFlag(value).successNel
+    case WdlString(value) if value.toLowerCase == "true" => ContinueOnReturnCodeFlag(true).successNel
+    case WdlString(value) if value.toLowerCase == "false" => ContinueOnReturnCodeFlag(false).successNel
+    case WdlInteger(value) => ContinueOnReturnCodeSet(Set(value)).successNel
+    case WdlArray(wdlType, seq) =>
+      val nels: Seq[ErrorOr[Int]] = seq map validateInt
+      val numFailures: Int = nels.count(_.isFailure)
+      if (numFailures == 0) {
+        val defaultReturnCodeNel = Set.empty[Int].successNel[String]
+        nels.foldLeft(defaultReturnCodeNel)((acc, v) => (acc |@| v) { (a, v) => a + v }) map ContinueOnReturnCodeSet
+      } else {
+        failureWithMessage
+      }
```

Comment:
@kshakir No problem, but I hope you won't mind if we start with some fundamentals as most of the functional programming languages basically use the same logic in their approach, and understanding those usually clears up most the confusion.  The rest is just syntactic sugar :)

So as the Computer Science department in the 1960s usually was part of the Mathematics department, a lot of this is based on certain foundations of mathematics which you already noticed.  So I will start with some definitions which I will slowly build up:

#### Symbols and Numbers (Atoms)

For simplicity, the smallest form of representation in any language would a symbol or number.  Numbers usually reside in a space (i.e. set of real number).

#### Operators

An operator is a symbol that maps from one space to another, and to simplify we map from the real number system to the real number system.  For example

Plus operator `(+)` maps the list `(1 2)` the number `3`.

Operators come from a branch of mathematics called [Operator theory](https://en.wikipedia.org/wiki/Operator_theory).

#### Functions

A function is a relation of inputs operated on, to provide an output.

For example:

  `x -> f(x)` is a function, where `x` is the input and `f(x)` is the function operating on it.  The inputs and outputs would be symbols.

#### Expressions

So an expression is a combination of symbols that has a syntax that can be used evaluation by an interpreter.  

Or more precisely:

```
   <expression> = <symbol> | <function>`
```
The `|` would be denote either (OR).

#### S-Expressions

These are symbolic expressions (s-exprs), and we can take the simplest form in languages such as LISP where we define as two types:

 - atom (i.e. a number, a character, a string, etc.)
 - list (a list of numbers, list of strings, list of combinations, etc.)

 Example:

```
  (1 2)
  (a b c)
  ( 1, "fluffy", "cat", "tail")
```
You can think of lists as sets from mathematics.

#### Functional Languages

In functional languages everything has a value.  Thus a function is constructed with operators (in prefix form) acting on values to produce values.  For example:

````
  (+ 1 2)
```
The operator + (prefixing the rest of the list) runs across the rest of the list to produce the value 3.  The reason why it is like this is because it naturally follows the mathematical interpretation of saying:

  ```
Add 1 with 2
```

#### [Lambda Calculus](https://en.wikipedia.org/wiki/Lambda_calculus)

[Lambda Calculus](https://en.wikipedia.org/wiki/Lambda_calculus) was created by [Alonzo Church](https://en.wikipedia.org/wiki/Alonzo_Church) in the 1930s to define the simplest way of formalizing a function with which computation can be performed - using mathematical logic - through the following form

```
   λ<symbolic_identifier>.<expression>
```

So let's take a simple function

   `λx.f[x]` which means a function of the form `x → f[x]` (i.e. process dependent variable `x` using the expression `f[x]`)

Below an example of such a function:

```
   λx.x^2
```

Now let's provide an input to it:

```
   (λx.x^2)4
```

The above would equal `16`.

Now we can even define a lambda expression as follows

```
  (lambda (parameters) body)
```

This is a list with the operator `lambda` processing the elements of the the list `(parameters)` using `body`.

Example:

```
  ((lambda (x y) (+ x y)) 2 3)
```

which will equal `5`.

In Scheme, which is a dialect of LISP you can use an operator called `define` to alias a function via a name to use later with custom values.

```
  (define square (lambda (x) (* x x)))
```

But we can then run a function over a list of number using another operator called `map`:

```
  (map square (list 1 2 3 4 5))
```

This would give me the following result:

```
  (1 4 9 16 25)
```

We can even reduce (by accumulating) through `foldl` (fold-left) operator:

```
  (foldl + 0 (list 1 2 3 4))
```

It takes the value 0 and then adds 1 and proceeds until the end of the list, which eventually gives us `(0 + (1 + (2 + (3 + (4 + 5)))))  = 10`.

  Thus now you can see how you can move functions over data, and next I'll cover how to provide context to functions (think of them as methods in class objects that would be reflective of the class - basically the inspection of the type of class).  This way you can reference functions, and operate functions with functions.

#### Functors, Applicative Functors (Applicatives) and Monads

So functors come from a branch of mathematics called [Category Theory](https://en.wikipedia.org/wiki/Category_theory), where if you let's say have two categories, A and B, a functor F maps objects in C to objects in D.  In terms of programming languages you can think of it as a type.  You have a function that can take an object of type A and map it as an object type B.

Applicative functors would be functions that are aware of the context (types) they are able to process.

Monads basically change the context (type) of the return type.  Think of Scala having Option where it can be None (i.e. empty) or Some (having some value).

#### Abstract vs Concrete

Now if a type is abstract it requires definitions or would not be able to be processed.  It can be referenced but not used like a concrete type.  The applicative functors require the concrete context of how to act.  If there is no definition, then they are lost on how to go forward.

Hope this overview helps in elucidating any worries/confusions/etc.  Feel free to ask if anything requires more clarification, as I would be glad to provide them.

My philosophy is don't be afraid to ask anything of anyone, no matter how insignificant it might seem to anyone.  I'm always in favor of as many folks to know as much as possible, and to have them feel they have a grasp of the subject if they want to.  I'm always pleasantly surprised at what folks can do once they know more ;)

Cheers,
Paul (pgrosu@gmail.com)